Translated 2020 CSS chapter to Italian (#2062)
HTTPArchive/almanac.httparchive.org: src/content/it/2020/css.md

Per alcune metriche, guardare al CSS AST non era sufficiente. Volevamo esaminare [SCSS](https://sass-lang.com/) ovunque fosse fornito tramite le mappe dei sorgenti poiché ci mostra ciò di cui gli sviluppatori hanno bisogno dai CSS che non è ancora possibile, mentre lo studio dei CSS ci mostra ciò che gli sviluppatori usano attualmente. Per questo, abbiamo dovuto utilizzare una *custom metric* — codice JavaScript che viene eseguito nel crawler quando visita una determinata pagina. Non è stato possibile utilizzare un parser SCSS appropriato in quanto potrebbe rallentare troppo la scansione, quindi abbiamo dovuto ricorrere a [espressioni regolari](https://github.com/LeaVerou/css-almanac/blob/master/runtime/sass.js). Nonostante l'approccio rozzo, abbiamo ottenuto [una sovrabbondanza di intuizioni](#sass)!

> Le metriche personalizzate (custom metrics) sono state utilizzate anche per parte dell'[analisi delle proprietà personalizzate](#custom-properties). Sebbene possiamo ottenere molte informazioni sull'utilizzo delle proprietà personalizzate solo dai fogli di stile, non possiamo costruire un grafico delle dipendenze senza essere in grado di guardare l'albero DOM per il contesto, poiché le proprietà personalizzate vengono ereditate. Guardando lo stile di calcolo dei nodi DOM ci dà anche informazioni come a quali tipi di elementi è applicata ciascuna proprietà e quali di essi sono [registrati](https://developer.mozilla.org/it/docs/Web/API/CSS/RegisterProperty) - informazioni che non possiamo ottenere dai fogli di stile.

Reviewer: Italian version doesn't exist:

```suggestion
Le metriche personalizzate (custom metrics) sono state utilizzate anche per parte dell'[analisi delle proprietà personalizzate](#custom-properties). Sebbene possiamo ottenere molte informazioni sull'utilizzo delle proprietà personalizzate solo dai fogli di stile, non possiamo costruire un grafico delle dipendenze senza essere in grado di guardare l'albero DOM per il contesto, poiché le proprietà personalizzate vengono ereditate. Guardando lo stile di calcolo dei nodi DOM ci dà anche informazioni come a quali tipi di elementi è applicata ciascuna proprietà e quali di essi sono [registrati](https://developer.mozilla.org/en-US/docs/Web/API/CSS/RegisterProperty) - informazioni che non possiamo ottenere dai fogli di stile.
```

  chart_url="https://docs.google.com/spreadsheets/d/e/2PACX-1vRpe_HsNGpekn6YZV9k6QGmcZPxalqnDrL7DrDY-7X65RZEf_-aGfWuEvhk-yWV83ctIceE1bppCLpj/pubchart?oid=1187401149&format=interactive",
  sheets_gid="863628849",
  sql_file="top_selector_classes_wp_fa_prefixes.sql"
) }}

### Gli ID

Nonostante gli ID siano scoraggiati in questi giorni in alcuni circoli a causa della loro specificità molto più elevata, la maggior parte dei siti Web li utilizza ancora, anche se con parsimonia. Meno della metà delle pagine utilizzava più di un ID in uno qualsiasi dei loro selettori (aveva una specificità massima di (1,x,y) o inferiore) e quasi tutte avevano una specificità mediana che non includeva ID (0,x,y) . Vedere la [specifica dei selettori](https://www.w3.org/TR/selectors/#specificity-rules) per maggiori dettagli sul calcolo della specificità e di questa notazione (a,b,c).

> Ma a cosa servono questi ID? Risulta che gli ID più popolari sono strutturali: `#content`, `#footer`, `#header`, `#main`, nonostante [gli elementi HTML corrispondenti](https://developer.mozilla.org/it/docs/Learn/HTML/Introduction_to_HTML/Document_and_website_structure#HTML_layout_elements_in_more_detail) esistenti che potrebbero essere utilizzati come selettori migliorando allo stesso tempo il markup semantico.

Reviewer: ```suggestion
Ma a cosa servono questi ID? Risulta che gli ID più popolari sono strutturali: `#content`, `#footer`, `#header`, `#main`, nonostante [gli elementi HTML corrispondenti](https://developer.mozilla.org/en-US/docs/Learn/HTML/Introduction_to_HTML/Document_and_website_structure#HTML_layout_elements_in_more_detail) esistenti che potrebbero essere utilizzati come selettori migliorando allo stesso tempo il markup semantico.
```

  sheets_gid="1743048352",
  sql_file="meta_important_adoption.sql"
) }}

Cos'è che gli sviluppatori sono così desiderosi di sovrascrivere? Abbiamo esaminato la suddivisione per proprietà e abbiamo scoperto che quasi l'80% delle pagine utilizza `!important` con la proprietà `display`. È una strategia comune applicare `display: none !important` per nascondere il contenuto nelle classi helper per sovrascrivere i CSS esistenti che usano `display` per definire una modalità di layout. Questo è un effetto collaterale di quello che, col senno di poi, era un difetto nei CSS. Ha combinato tre funzionalità ortogonali in una: la modalità di layout interno, il comportamento del flusso e lo stato di visibilità sono tutti controllati dalla proprietà `display`. Sono stati fatti sforzi per separare questi valori in singole parole chiave `display` in modo che possano essere regolate individualmente tramite proprietà personalizzate, ma per il momento [il supporto del browser è praticamente inesistente.](https://caniuse.com/mdn-css_properties_display_multi-keyword_values)

{{ figure_markup(
  image="important-top-properties.png",

>   alt="The top !important properties by the percent of pages.",

Reviewer: Missed a translation here

  description="Grafico a barre che mostra la percentuale di pagine che utilizzano la sintassi della pseudo-class (preceduta da un due punti) rispetto alla sintassi degli pseudo-element (due punti) per gli pseudo-elements. Lo pseudo-element `before` viene utilizzato con la sintassi della pseudo-class sul 71% delle pagine mobile e la sintassi dello pseudo-element sul 33% delle pagine mobile. Lo pseudo-element `after` viene utilizzato con la sintassi di classe ed elemento sul 68% e 30% delle pagine mobile,`first-letter` sul 7% e 1% delle pagine mobile e `first-line` sull'1% e 0 % di pagine per dispositivi mobile.",
  chart_url="https://docs.google.com/spreadsheets/d/e/2PACX-1vRpe_HsNGpekn6YZV9k6QGmcZPxalqnDrL7DrDY-7X65RZEf_-aGfWuEvhk-yWV83ctIceE1bppCLpj/pubchart?oid=227968207&format=interactive",
  sheets_gid="2029589646",
  sql_file="top_selector_pseudo_classes.sql"
) }}

Le pseudo-classi di gran lunga più popolari sono quelle di azione dell'utente, con `:hover`, `:focus`, e `:active` all'inizio dell'elenco, tutte utilizzate in oltre due terzi delle pagine, a indicare che agli sviluppatori piace la comodità di specificare le interazioni dichiarative dell'interfaccia utente.

> `:root` sembra molto più popolare di quanto sia giustificato dalla sua funzione, usata in un terzo delle pagine. Nel contenuto HTML, seleziona semplicemente l'elemento `<html>`, quindi perché gli sviluppatori non hanno usato semplicemente `html`? Una possibile risposta potrebbe risiedere in una pratica comune relativa alla definizione di proprietà personalizzate, [che sono anche molto usate](#custom-properties), sulla pseudo-class `:root`. Un'altra risposta potrebbe risiedere nella specificità: `:root`, essendo una pseudo-classe, ha una specificità maggiore di `html`: (0,1,0) vs (0,0,1). È un trucco comune aumentare la specificità di un selettore anteponendolo a `:root`, ad es. `:root .foo` ha una specificità di (0, 2, 0) rispetto a solo (0, 1, 0) per `.foo`. Questo è spesso tutto ciò che serve per spingere un selettore leggermente sopra un altro nella cascade race ed evitare la martellata che è `!important`. Per verificare questa ipotesi, abbiamo anche misurato esattamente questo: quante pagine usano `:root` all'inizio di un selettore discendente? I risultati hanno verificato la nostra ipotesi: un notevole 29% delle pagine usa `:root` in questo modo! Inoltre, il 14% delle pagine desktop e il 19% delle pagine mobile utilizzano `html` all'inizio di un selettore discendente, forse per dare al selettore un incremento di specificità ancora minore. La popolarità di questi hack di specificità indica fortemente che gli sviluppatori necessitano di un controllo granulare più fine per modificare la specificità rispetto a quello che viene loro offerto tramite `!important`. Per fortuna, arriverà presto con [`:where()`](https://developer.mozilla.org/it/docs/Web/CSS/:where), che è già [implementato su tutta la linea](https://caniuse.com/mdn-css_selectors_where) (anche se dietro una bandiera in Chrome per ora).

Reviewer: ```suggestion
`:root` sembra molto più popolare di quanto sia giustificato dalla sua funzione, usata in un terzo delle pagine. Nel contenuto HTML, seleziona semplicemente l'elemento `<html>`, quindi perché gli sviluppatori non hanno usato semplicemente `html`? Una possibile risposta potrebbe risiedere in una pratica comune relativa alla definizione di proprietà personalizzate, [che sono anche molto usate](#custom-properties), sulla pseudo-class `:root`. Un'altra risposta potrebbe risiedere nella specificità: `:root`, essendo una pseudo-classe, ha una specificità maggiore di `html`: (0,1,0) vs (0,0,1). È un trucco comune aumentare la specificità di un selettore anteponendolo a `:root`, ad es. `:root .foo` ha una specificità di (0, 2, 0) rispetto a solo (0, 1, 0) per `.foo`. Questo è spesso tutto ciò che serve per spingere un selettore leggermente sopra un altro nella cascade race ed evitare la martellata che è `!important`. Per verificare questa ipotesi, abbiamo anche misurato esattamente questo: quante pagine usano `:root` all'inizio di un selettore discendente? I risultati hanno verificato la nostra ipotesi: un notevole 29% delle pagine usa `:root` in questo modo! Inoltre, il 14% delle pagine desktop e il 19% delle pagine mobile utilizzano `html` all'inizio di un selettore discendente, forse per dare al selettore un incremento di specificità ancora minore. La popolarità di questi hack di specificità indica fortemente che gli sviluppatori necessitano di un controllo granulare più fine per modificare la specificità rispetto a quello che viene loro offerto tramite `!important`. Per fortuna, arriverà presto con [`:where()`](https://developer.mozilla.org/en-US/docs/Web/CSS/:where), che è già [implementato su tutta la linea](https://caniuse.com/mdn-css_selectors_where) (anche se dietro una bandiera in Chrome per ora).
```

  chart_url="https://docs.google.com/spreadsheets/d/e/2PACX-1vRpe_HsNGpekn6YZV9k6QGmcZPxalqnDrL7DrDY-7X65RZEf_-aGfWuEvhk-yWV83ctIceE1bppCLpj/pubchart?oid=1363774711&format=interactive",
  sheets_gid="2029589646",
  sql_file="top_selector_pseudo_classes.sql"
) }}

Quando si tratta di pseudo-element, dopo i soliti sospetti `::before` e `::after`, quasi tutti gli pseudo-element popolari erano estensioni del browser per lo stile dei controlli dei form e altre UI incorporate, facendo fortemente eco alla necessità degli sviluppatori di un controllo più dettagliato sullo stile dell'interfaccia utente incorporata. Lo stile di focus rings, placeholders, search inputs, spinners, selection, scrollbars, media controls era particolarmente popolare.

{{ figure_markup(
  image="popular-selector-pseudo-elements.png",

>   caption="Grafico a barre che mostra gli pseudo-element più popolari come percentuale delle pagine per desktop e dispositivi mobile. Desktop e dispositivi mobili sono per lo più simili, con i dispositivi mobili che tendono ad avere un'adozione leggermente maggiore. Lo pseudo-element più popolare è `before`, utilizzato sul 33% delle pagine per dispositivi mobile. Lo pseudo-element `after` viene utilizzato nel 30% delle pagine mobile. `-moz-focus-inner` viene utilizzato nel 24% delle pagine. La popolarità scende dopo quelli dal 17% al 4% in ordine decrescente: `-webkit-input-placeholder`, `-moz-placeholder`, `-webkit-search-decoration`, `-webkit-search-cancel-button`, `-webkit-inner-spin-button`, `-webkit-outer-spin-button`, `-webkit-scrollbar` (7%), `selection`, `-ms-clear`, `-moz-selection`, `-webkit-media-controls`, e `-webkit-scrollbar-thumb`.",

Reviewer: ```suggestion
  caption="Le pseudo-elements più popolari come percentuale di pagine.",
  description="Grafico a barre che mostra gli pseudo-element più popolari come percentuale delle pagine per desktop e dispositivi mobile. Desktop e dispositivi mobili sono per lo più simili, con i dispositivi mobili che tendono ad avere un'adozione leggermente maggiore. Lo pseudo-element più popolare è `before`, utilizzato sul 33% delle pagine per dispositivi mobile. Lo pseudo-element `after` viene utilizzato nel 30% delle pagine mobile. `-moz-focus-inner` viene utilizzato nel 24% delle pagine. La popolarità scende dopo quelli dal 17% al 4% in ordine decrescente: `-webkit-input-placeholder`, `-moz-placeholder`, `-webkit-search-decoration`, `-webkit-search-cancel-button`, `-webkit-inner-spin-button`, `-webkit-outer-spin-button`, `-webkit-scrollbar` (7%), `selection`, `-ms-clear`, `-moz-selection`, `-webkit-media-controls`, e `-webkit-scrollbar-thumb`.",
```

  chart_url="https://docs.google.com/spreadsheets/d/e/2PACX-1vRpe_HsNGpekn6YZV9k6QGmcZPxalqnDrL7DrDY-7X65RZEf_-aGfWuEvhk-yWV83ctIceE1bppCLpj/pubchart?oid=166603845&format=interactive",
  sheets_gid="1221511608",
  sql_file="units_frequency.sql"
) }}

Le lunghezze sono gli unici tipi di valori CSS per i quali possiamo omettere l'unità quando il valore è zero, cioè possiamo scrivere `0` invece di `0px` o `0em` ecc. Gli sviluppatori (o CSS minificatori?) ne fanno un ampio uso: di tutti i valori `0`, l'89% era senza unità.

{{ figure_markup(
  image="zero-lengths.png",
  caption="Popolarità relativa di 0 lunghezze per unità come percentuale di occorrenze sulle pagine per dispositivi mobile.",
  description="Grafico a torta che mostra 0 lunghezze senza unità (AKA senza unità) utilizzato l'88.7% delle volte su pagine mobile, l'unità `px` al 10.7% e altre unità allo 0.5% delle istanze.",
  chart_url="https://docs.google.com/spreadsheets/d/e/2PACX-1vRpe_HsNGpekn6YZV9k6QGmcZPxalqnDrL7DrDY-7X65RZEf_-aGfWuEvhk-yWV83ctIceE1bppCLpj/pubchart?oid=1935151776&format=interactive",
  sheets_gid="313150061",
  sql_file="units_zero.sql"
) }}

### Calcoli

> Quando la funzione [`calc()`](https://developer.mozilla.org/it/docs/Web/CSS/calc()) è stata introdotta per eseguire calcoli tra diverse unità in CSS, è stata una rivoluzione. In precedenza, solo i preprocessori erano in grado di accogliere tali calcoli, ma i risultati erano limitati a valori statici e inaffidabili, poiché mancava il contesto dinamico che è spesso necessario.

Reviewer: ```suggestion
Quando la funzione [`calc()`](https://developer.mozilla.org/en-US/docs/Web/CSS/calc()) è stata introdotta per eseguire calcoli tra diverse unità in CSS, è stata una rivoluzione. In precedenza, solo i preprocessori erano in grado di accogliere tali calcoli, ma i risultati erano limitati a valori statici e inaffidabili, poiché mancava il contesto dinamico che è spesso necessario.
```

  sheets_gid="2077258816",
  sql_file="calc_operators.sql"
) }}

La maggior parte dei calcoli è molto semplice, con il 99.5% dei calcoli che coinvolgono fino a 2 unità diverse, l'88.5% dei calcoli che coinvolgono fino a 2 operatori e il 99.4% dei calcoli che coinvolgono una serie di parentesi o meno (3 calcoli su 4 non includono affatto parentesi ).

{{ figure_markup(
  image="calc-complexity-units.png",
  alt="Distribuzione del numero di unità per occorrenza calc().",
  caption="Distribuzione del numero di unità per occorrenza `calc()`.",
  description="Grafico a barre che mostra la distribuzione del numero di unità per occorrenza della funzione di calc. Desktop e mobile hanno risultati simili. Calc viene utilizzato con un'unità per l'11% delle volte sulle pagine mobile, due volte l'89% delle volte e 3 o più volte circa lo 0% delle volte.",
  chart_url="https://docs.google.com/spreadsheets/d/e/2PACX-1vRpe_HsNGpekn6YZV9k6QGmcZPxalqnDrL7DrDY-7X65RZEf_-aGfWuEvhk-yWV83ctIceE1bppCLpj/pubchart?oid=695698141&format=interactive",
  sheets_gid="1493602565",
  sql_file="calc_complexity_units.sql"
)
}}

### Parole chiave globali e `all`

> Per molto tempo, CSS ha supportato solo una parola chiave globale: [`inherit`](https://developer.mozilla.org/it/docs/Web/CSS/inherit), che consente il ripristino di una proprietà ereditabile per il suo valore ereditato o riutilizzando il valore del genitore per una data proprietà non ereditabile. Si scopre che il primo è molto più comune del secondo, con l'81.37% dell'utilizzo di `inherit` trovato su proprietà ereditabili. Il resto è principalmente ereditare sfondi, bordi o dimensioni. Quest'ultimo probabilmente indica problemi di layout, poiché con la giusta modalità di layout uno raramente ha bisogno di forzare `width` e `height` da ereditare.

Reviewer: ```suggestion
Per molto tempo, CSS ha supportato solo una parola chiave globale: [`inherit`](https://developer.mozilla.org/en-US/docs/Web/CSS/inherit), che consente il ripristino di una proprietà ereditabile per il suo valore ereditato o riutilizzando il valore del genitore per una data proprietà non ereditabile. Si scopre che il primo è molto più comune del secondo, con l'81.37% dell'utilizzo di `inherit` trovato su proprietà ereditabili. Il resto è principalmente ereditare sfondi, bordi o dimensioni. Quest'ultimo probabilmente indica problemi di layout, poiché con la giusta modalità di layout uno raramente ha bisogno di forzare `width` e `height` da ereditare.
```

La parola chiave `inherit` è particolarmente utile per reimpostare il colore del link predefinito con il colore del testo principale. Non sorprende quindi che `color` sia la proprietà più comune su cui viene usata `inherit`. Quasi un terzo di tutto l'utilizzo di `inherit` si trova nella proprietà `color`. Il 75% delle pagine usa `color: inherit` almeno una volta.

> Sebbene il *valore iniziale* di una proprietà sia un concetto che [esiste dai CSS 1](https://www.w3.org/TR/CSS1/#cascading-order), ha solo la sua parola chiave dedicata, `initial`, per fare riferimento esplicito ad esso [17 anni dopo](https://www.w3.org/TR/2013/WD-css3-cascade-20130103/#initial-keyword) e ci sono voluti altri due anni per quella parola chiave ottenere [supporto browser universale](https://caniuse.com/css-initial-value) nel 2015. Non sorprende quindi che sia usato molto meno di `inherit`. Mentre la vecchia inherit si trova nell'85% delle pagine, `initial` appare nel 51% di esse. Inoltre, c'è molta confusione su cosa fa effettivamente `initial`, dal momento che `display` è in cima alla lista delle proprietà più comunemente usate con `initial`, con `display: initial` che appare nel 10% delle pagine. Presumibilmente, gli sviluppatori hanno pensato che questo resetta `display` al suo valore dal valore [user agent stylesheet](https://developer.mozilla.org/it/docs/Web/CSS/Cascade#User-agent_stylesheets) e lo usavamo per attivare e disattivare `display: none`. Tuttavia, [il valore iniziale di `display` è `inline`](https://drafts.csswg.org/css-display/#the-display-properties), quindi `display: initial` è solo un altro modo di scrivere `display: inline` e non ha proprietà magiche dipendenti dal contesto.

Reviewer: ```suggestion
Sebbene il *valore iniziale* di una proprietà sia un concetto che [esiste dai CSS 1](https://www.w3.org/TR/CSS1/#cascading-order), ha solo la sua parola chiave dedicata, `initial`, per fare riferimento esplicito ad esso [17 anni dopo](https://www.w3.org/TR/2013/WD-css3-cascade-20130103/#initial-keyword) e ci sono voluti altri due anni per quella parola chiave ottenere [supporto browser universale](https://caniuse.com/css-initial-value) nel 2015. Non sorprende quindi che sia usato molto meno di `inherit`. Mentre la vecchia inherit si trova nell'85% delle pagine, `initial` appare nel 51% di esse. Inoltre, c'è molta confusione su cosa fa effettivamente `initial`, dal momento che `display` è in cima alla lista delle proprietà più comunemente usate con `initial`, con `display: initial` che appare nel 10% delle pagine. Presumibilmente, gli sviluppatori hanno pensato che questo resetta `display` al suo valore dal valore [user agent stylesheet](https://developer.mozilla.org/en-US/docs/Web/CSS/Cascade#User-agent_stylesheets) e lo usavamo per attivare e disattivare `display: none`. Tuttavia, [il valore iniziale di `display` è `inline`](https://drafts.csswg.org/css-display/#the-display-properties), quindi `display: initial` è solo un altro modo di scrivere `display: inline` e non ha proprietà magiche dipendenti dal contesto.
```

Sembra che ci siano due ragioni per cui il web è ancora intrappolato nella terra sRGB. Il primo è la mancanza di strumenti, la mancanza di buoni selettori di colori, la mancanza di comprensione di quali colori più vividi siano disponibili. Ma il motivo principale, pensiamo, è che fino ad oggi Safari è l'unico browser a implementarlo. La situazione sta cambiando rapidamente - Chrome e Firefox stanno entrambi implementando in questo momento - ma fino a quando non verrà fornito il supporto, probabilmente utilizzare display-p3 è uno sforzo eccessivo per un guadagno insufficiente perché [solo il 17% degli spettatori](https://gs.statcounter.com/browser-market-share) vedrà quei colori. La maggior parte delle persone vedrà il fallback. Quindi l'uso attuale è un sottile cambiamento nella vivacità del colore, piuttosto che una grande differenza.

Sarà interessante vedere come l'uso del colore display-p3 (esistono altre opzioni, ma questa è l'unica che abbiamo trovato sul web) cambia nel corso del prossimo anno o due.

> Perché *l'ampia gamma di colori* (wide color gamut o brevemente WCG) è solo l'inizio. L'industria televisiva e cinematografica è già passata dalla P3 a una gamma ancora più ampia, [*Rec. 2020*](https://en.wikipedia.org/wiki/Rec._2020); e anche una gamma più ampia di luminosità, dai riflessi accecanti alle ombre più profonde. *High Dynamic Range* (HDR) è già arrivato nelle nostre case, soprattutto su giochi, TV in streaming e film. Il web ha molto da recuperare.

Reviewer: ```suggestion
Perché *l'ampia gamma di colori* (wide color gamut o brevemente WCG) è solo l'inizio. L'industria televisiva e cinematografica è già passata dalla P3 a una gamma ancora più ampia, [*Rec. 2020*](https://it.wikipedia.org/wiki/BT.2020); e anche una gamma più ampia di luminosità, dai riflessi accecanti alle ombre più profonde. *High Dynamic Range* (HDR) è già arrivato nelle nostre case, soprattutto su giochi, TV in streaming e film. Il web ha molto da recuperare.
```

Il gradiente con il maggior numero di interruzioni di colore è [questo](https://dabblet.com/gist/4d1637d78c71ef2d8d37952fc6e90ff5) con 646 interruzioni! Questo è quasi certamente generato e il codice CSS risultante è 8KB, quindi un PNG alto 1px avrebbe probabilmente fatto anche il lavoro, con un'impronta più piccola (la nostra immagine sotto è di 1,1 KB).

{{ figure_markup(
  image="gradient-most-stops.png",
  classes="height-16vw-122px",
  caption="Il gradiente con il maggior numero di interruzioni di colore, 646.",
  description="Uno screenshot del gradiente con il maggior numero di interruzioni di colore, che è una serie di intricate strisce multicolori di varie tonalità.",
  alt="Uno screenshot del gradiente con il maggior numero di interruzioni di colore, che è una serie di intricate strisce multicolori di varie tonalità.",
  width=600,
  height=122
) }}

## Layout

CSS ora ha una serie di opzioni di layout che lo distinguono dalle tabelle che dovevano essere utilizzate per il layout. I layout Flexbox, Grid e Multiple-column sono supportati dalla maggior parte dei browser.

### Adozione di Flexbox e Grid

> [L'edizione 2019](../2019/css#flexbox), ha riportato che il 41% delle pagine su dispositivi mobile e desktop conteneva proprietà [Flexbox](https://developer.mozilla.org/it/docs/Web/CSS/CSS_Flexible_Box_Layout/Basic_Concepts_of_Flexbox). Nel 2020, questo numero è cresciuto al 63% per i dispositivi mobile e al 65% per i desktop. Questo metodo di layout è ampiamente adottato, poiché esistono ancora numerosi siti legacy sviluppati prima che Flexbox diventasse uno strumento valido.
> 
> Se guardiamo [Grid layout](https://developer.mozilla.org/it/docs/Web/CSS/CSS_Grid_Layout), la percentuale di siti che utilizzano il layout Grid è aumentata al 4% per i dispositivi mobile e al 5% per desktop. L'utilizzo è raddoppiato dall'anno scorso, ma è ancora molto indietro rispetto al layout flessibile.

Reviewer: ```suggestion
[L'edizione 2019](../2019/css#flexbox), ha riportato che il 41% delle pagine su dispositivi mobile e desktop conteneva proprietà [Flexbox](https://developer.mozilla.org/it/docs/Web/CSS/CSS_Flexible_Box_Layout/Basic_Concepts_of_Flexbox). Nel 2020, questo numero è cresciuto al 63% per i dispositivi mobile e al 65% per i desktop. Questo metodo di layout è ampiamente adottato, poiché esistono ancora numerosi siti legacy sviluppati prima che Flexbox diventasse uno strumento valido.

Se guardiamo [Grid layout](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_Grid_Layout), la percentuale di siti che utilizzano il layout Grid è aumentata al 4% per i dispositivi mobile e al 5% per desktop. L'utilizzo è raddoppiato dall'anno scorso, ma è ancora molto indietro rispetto al layout flessibile.
```

  sheets_gid="335708969",
  sql_file="layout_properties.sql"
) }}

Dato che Flexbox era utilizzabile nei browser prima del layout Grid, è probabile che parte dell'utilizzo di Flexbox sia per la configurazione di un grid system. Per utilizzare Flexbox come grid, gli autori devono disabilitare parte della flessibilità intrinseca di Flexbox. Per fare ciò, imposta la proprietà `flex-grow` su `0`, quindi dimensiona gli elementi flessibili usando le percentuali. Utilizzando queste informazioni siamo stati in grado di segnalare che il 19% dei siti sia su desktop che su dispositivi mobile utilizzava Flexbox in questo modo simile a una griglia.

I motivi per scegliere Flexbox su Grid sono spesso citati come supporto del browser, dato che il layout Grid [non era supportato in Internet Explorer](https://caniuse.com/css-grid). Inoltre, alcuni autori potrebbero non aver ancora imparato il layout Grid o utilizzare un framework con un sistema grid basato su Flexbox. Il framework [Bootstrap](https://getbootstrap.com/docs/4.5/layout/grid/) attualmente utilizza un grid basato su Flexbox, che è comune con alcune scelte di framework popolari.

### Utilizzo di diverse tecniche di layout Grid

> La specifica del layout Grid offre diversi modi per descrivere e definire il layout in CSS. L'utilizzo più elementare prevede la disposizione degli elementi [da una grid line all'altra](https://www.smashingmagazine.com/2020/01/understanding-css-grid-lines/). Che dire delle [linee di denominazione](https://developer.mozilla.org/it/docs/Web/CSS/CSS_Grid_Layout/Layout_using_Named_Grid_Lines) o dell'uso di `grid-template-areas`?

Reviewer: ```suggestion
La specifica del layout Grid offre diversi modi per descrivere e definire il layout in CSS. L'utilizzo più elementare prevede la disposizione degli elementi [da una grid line all'altra](https://www.smashingmagazine.com/2020/01/understanding-css-grid-lines/). Che dire delle [linee di denominazione](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_Grid_Layout/Layout_using_Named_Grid_Lines) o dell'uso di `grid-template-areas`?
```

Per le righe con nome, abbiamo verificato la presenza di parentesi quadre in un elenco di tracce. Il nome o i nomi inseriti tra parentesi quadre.

```css
.wrapper {
  display: grid;
  grid-template-columns: [main-start] 1fr [content-start] 1fr [content-end] 1fr [main-end];
}
```

Di conseguenza, lo 0.23% delle pagine che utilizzano Grid sui dispositivi mobile avevano linee denominate, rispetto allo 0.27% sui desktop.

> La funzione [Grid template areas](https://developer.mozilla.org/it/docs/Web/CSS/CSS_Grid_Layout/Grid_Template_Areas), che consente agli autori di denominare gli grid item e di posizionarli sulla griglia come valore della proprietà `grid-template-areas`. Dei siti che utilizzano Grid, il 19% su dispositivi mobile e il 20% su desktop utilizzava questo metodo.

Reviewer: ```suggestion
La funzione [Grid template areas](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_Grid_Layout/Grid_Template_Areas), che consente agli autori di denominare gli grid item e di posizionarli sulla griglia come valore della proprietà `grid-template-areas`. Dei siti che utilizzano Grid, il 19% su dispositivi mobile e il 20% su desktop utilizzava questo metodo.
```

Questi risultati mostrano che non solo l'utilizzo del layout Grid è ancora relativamente basso sui siti Web di produzione, ma anche il suo utilizzo è relativamente semplice. L'autore preferisce una semplice disposizione basata sulla riga rispetto al modo in cui è possibile denominare righe e aree. Sebbene non ci sia nulla di sbagliato nella scelta di farlo, mi chiedo se la lenta adozione del layout Grid sia in parte dovuta al fatto che gli autori non hanno ancora realizzato la potenza di queste funzionalità. Se il layout Grid è visto essenzialmente come Flexbox con uno scarso supporto del browser, questo lo renderebbe sicuramente una scelta meno convincente.

### Layout a più colonne

> Il [layout a più colonne](https://developer.mozilla.org/it/docs/Web/CSS/CSS_Columns/Basic_Concepts_of_Multicol) o *multicol*, consente la disposizione del contenuto in colonne, proprio come un quotidiano. Sebbene popolare nei CSS come utilizzato per la stampa, è meno utile sul Web a causa del rischio di creare una situazione in cui un lettore deve scorrere su e giù per leggere il contenuto. Sulla base dei dati, tuttavia, ci sono molte più pagine che utilizzano il layout multicol rispetto al layout Grid con il 15.33% su desktop e il 14.95% su dispositivi mobile. Sebbene le proprietà base multicol siano ben supportate, l'uso più complesso e il controllo delle column break con [fragmentation](https://www.smashingmagazine.com/2019/02/css-fragmentation/) ha [supporto irregolare](https://caniuse.com/multicolumn). Detto questo, è stato abbastanza sorprendente che ci fossero così tanti usi.

Reviewer: ```suggestion
Il [layout a più colonne](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_Columns/Basic_Concepts_of_Multicol) o *multicol*, consente la disposizione del contenuto in colonne, proprio come un quotidiano. Sebbene popolare nei CSS come utilizzato per la stampa, è meno utile sul Web a causa del rischio di creare una situazione in cui un lettore deve scorrere su e giù per leggere il contenuto. Sulla base dei dati, tuttavia, ci sono molte più pagine che utilizzano il layout multicol rispetto al layout Grid con il 15.33% su desktop e il 14.95% su dispositivi mobile. Sebbene le proprietà base multicol siano ben supportate, l'uso più complesso e il controllo delle column break con [fragmentation](https://www.smashingmagazine.com/2019/02/css-fragmentation/) ha [supporto irregolare](https://caniuse.com/multicolumn). Detto questo, è stato abbastanza sorprendente che ci fossero così tanti usi.
```

### Box sizing

> È utile sapere quanto saranno grandi le caselle sulla tua pagina, ma con il [standard CSS box model](https://developer.mozilla.org/it/docs/Learn/CSS/Building_blocks/The_box_model#What_is_the_CSS_box_model) aggiungendo padding e border alla dimensione della content-box, la dimensione che hai dato alla tua casella è più piccola della casella visualizzata sulla tua pagina. Sebbene non possiamo cambiare la cronologia, la proprietà `box-sizing` consente agli autori di passare ad applicare la dimensione specificata al `border-box`, quindi la dimensione che imposti è la dimensione che vedi renderizzata. Quanti siti usano la proprietà `box-sizing`? La maggior parte di loro! La proprietà `box-sizing` appare nell'83.79% dei CSS desktop e nell'86.39% sui dispositivi mobile.

Reviewer: ```suggestion
È utile sapere quanto saranno grandi le caselle sulla tua pagina, ma con il [standard CSS box model](https://developer.mozilla.org/en-US/docs/Learn/CSS/Building_blocks/The_box_model#What_is_the_CSS_box_model) aggiungendo padding e border alla dimensione della content-box, la dimensione che hai dato alla tua casella è più piccola della casella visualizzata sulla tua pagina. Sebbene non possiamo cambiare la cronologia, la proprietà `box-sizing` consente agli autori di passare ad applicare la dimensione specificata al `border-box`, quindi la dimensione che imposti è la dimensione che vedi renderizzata. Quanti siti usano la proprietà `box-sizing`? La maggior parte di loro! La proprietà `box-sizing` appare nell'83.79% dei CSS desktop e nell'86.39% sui dispositivi mobile.
```

  chart_url="https://docs.google.com/spreadsheets/d/e/2PACX-1vRpe_HsNGpekn6YZV9k6QGmcZPxalqnDrL7DrDY-7X65RZEf_-aGfWuEvhk-yWV83ctIceE1bppCLpj/pubchart?oid=709747830&format=interactive",
  sheets_gid="1998209374",
  sql_file="transition_animation_names.sql"
) }}

## Effetti visivi

CSS offre anche una grande varietà di effetti visivi che danno ai progettisti l'accesso a tecniche di progettazione avanzate integrate nei browser a cui è possibile accedere con piccole quantità di codice.

### I metodi di fusione

L'anno scorso, l'8% delle pagine utilizzava i metodi di fusione (blend modes). Quest'anno, l'adozione è aumentata in modo significativo, con il 13% delle pagine che utilizza modalità di fusione sugli elementi (`mix-blend-mode`) e il 2% sui background (`background-blend-mode`).

### Filtri

> L'adozione dei filtri è rimasta elevata, con la proprietà `filter` che compare nel 79.43% delle pagine. Anche se all'inizio era abbastanza eccitante, è probabile che molti siano vecchi [filtri DX IE](https://developer.mozilla.org/it/docs/Archive/Web/CSS/-ms-filter), che condividevano lo stesso nome di proprietà. Quando abbiamo preso in considerazione solo i filtri CSS validi che Blink riconosce, l'utilizzo scende al 22% per i dispositivi mobile e al 20% per i desktop, con `blur()` che è il tipo di filtro più popolare e compare nel 4% delle pagine.

Reviewer: ```suggestion
L'adozione dei filtri è rimasta elevata, con la proprietà `filter` che compare nel 79.43% delle pagine. Anche se all'inizio era abbastanza eccitante, è probabile che molti siano vecchi [filtri DX IE](https://developer.mozilla.org/en-US/docs/Archive/Web/CSS/-ms-filter), che condividevano lo stesso nome di proprietà. Quando abbiamo preso in considerazione solo i filtri CSS validi che Blink riconosce, l'utilizzo scende al 22% per i dispositivi mobile e al 20% per i desktop, con `blur()` che è il tipo di filtro più popolare e compare nel 4% delle pagine.
```

  chart_url="https://docs.google.com/spreadsheets/d/e/2PACX-1vRpe_HsNGpekn6YZV9k6QGmcZPxalqnDrL7DrDY-7X65RZEf_-aGfWuEvhk-yWV83ctIceE1bppCLpj/pubchart?oid=262191540&format=interactive",
  sheets_gid="1368222498",
  sql_file="custom_property_depth.sql"
) }}

L'esame dei selettori su cui vengono dichiarate le proprietà personalizzate conferma ulteriormente che la maggior parte dell'utilizzo di proprietà personalizzate naturali sono abbastanza semplici. Due delle tre dichiarazioni di proprietà personalizzate si trovano sull'elemento radice, a indicare che vengono utilizzate essenzialmente come costanti globali. È importante notare che molti polyfill popolari hanno richiesto che fossero globali in questo senso, quindi gli sviluppatori che utilizzano tali polyfill potrebbero non aver avuto scelta.

## CSS e JS

Negli ultimi anni, l'interazione tra CSS e JavaScript è cresciuta, non solo impostando e disattivando classi e stili CSS. Quindi quanto usi tecniche come Houdini e tecniche come CSS-in-JS?

### Houdini

> Probabilmente avrai già sentito parlare di [Houdini](https://developer.mozilla.org/it/docs/Web/Houdini). Houdini è un insieme di API di basso livello che espone parti del motore CSS, dando agli sviluppatori il potere di estendere CSS agganciandosi allo stile e al processo di layout del motore di rendering di un browser. Poiché [diverse specifiche Houdini sono state fornite nei browser](https://ishoudinireadyyet.com/), abbiamo pensato che fosse ora di vedere se sono state effettivamente utilizzate allo stato brado. Risposta breve: no. E ora per la risposta più lunga ...
> 
> In primo luogo, abbiamo esaminato le [API Properties e Values](https://developer.mozilla.org/it/docs/Web/API/CSS/RegisterProperty), che consente agli sviluppatori di registrare una proprietà personalizzata e assegnarle un type, un valore iniziale e impedire che venga ereditato. Uno dei casi d'uso principali è la possibilità di animare le proprietà personalizzate, quindi abbiamo anche esaminato la frequenza con cui le proprietà personalizzate vengono animate.

Reviewer: ```suggestion
Probabilmente avrai già sentito parlare di [Houdini](https://developer.mozilla.org/en-US/docs/Web/Houdini). Houdini è un insieme di API di basso livello che espone parti del motore CSS, dando agli sviluppatori il potere di estendere CSS agganciandosi allo stile e al processo di layout del motore di rendering di un browser. Poiché [diverse specifiche Houdini sono state fornite nei browser](https://ishoudinireadyyet.com/), abbiamo pensato che fosse ora di vedere se sono state effettivamente utilizzate allo stato brado. Risposta breve: no. E ora per la risposta più lunga ...

In primo luogo, abbiamo esaminato le [API Properties e Values](https://developer.mozilla.org/en-US/docs/Web/API/CSS/RegisterProperty), che consente agli sviluppatori di registrare una proprietà personalizzata e assegnarle un type, un valore iniziale e impedire che venga ereditato. Uno dei casi d'uso principali è la possibilità di animare le proprietà personalizzate, quindi abbiamo anche esaminato la frequenza con cui le proprietà personalizzate vengono animate.
```

Come spesso accade con la tecnologia più avanzata, specialmente quando non è supportata in tutti i browser, l'adozione naturale è stata estremamente bassa. È stato riscontrato che solo 32 pagine desktop e 20 pagine per dispositivi mobile avevano proprietà personalizzate registrate, sebbene ciò escluda le proprietà personalizzate registrate ma non applicate al momento della scansione. Solo 325 pagine per dispositivi mobile e 330 pagine desktop (0,00%) utilizzano proprietà personalizzate nelle animazioni e la maggior parte (74%) sembra essere guidata da un [componente Vue](https://quasar.dev/vue-components/expansion-item). Praticamente nessuno di questi sembra averli registrati, anche se è probabile che l'animazione non fosse attiva al momento della scansione, quindi non era necessario registrare alcun stile di calcolo.

> La [Paint API](https://developer.mozilla.org/it/docs/Web/API/CSS_Painting_API) è una specifica Houdini implementata in modo più ampio che consente agli sviluppatori di creare funzioni CSS personalizzate che restituiscono valori `<image>`, ad es per implementare sfumature o pattern personalizzati. È stato riscontrato che solo 12 pagine utilizzano `paint()`. È stato visualizzato un solo nome del worklet (`hexagon`, `ruler`, `lozenge`, `image-cross`, `grid`, `dashed-line`, `ripple`) per pagina

Reviewer: ```suggestion
La [Paint API](https://developer.mozilla.org/en-US/docs/Web/API/CSS_Painting_API) è una specifica Houdini implementata in modo più ampio che consente agli sviluppatori di creare funzioni CSS personalizzate che restituiscono valori `<image>`, ad es per implementare sfumature o pattern personalizzati. È stato riscontrato che solo 12 pagine utilizzano `paint()`. È stato visualizzato un solo nome del worklet (`hexagon`, `ruler`, `lozenge`, `image-cross`, `grid`, `dashed-line`, `ripple`) per pagina
```

### Proprietà logiche e fisiche

Molte delle prime proprietà che ci vengono insegnate quando impariamo i CSS, cose come `width`, `height`, `margin-left`,`padding-bottom`, `right` e così via sono basate su una specifica direzione fisica. Tuttavia, quando il contenuto deve essere presentato in più lingue con diverse caratteristiche di direzionalità, queste direzioni fisiche dipendono spesso dalla lingua, ad es. `margin-left` spesso ha bisogno di diventare `margin-right` in una lingua da destra a sinistra come l'arabo. La direzionalità è una caratteristica 2D. Ad esempio, `height` potrebbe dover diventare `width` quando presentiamo il contenuto in scrittura verticale (come il cinese tradizionale).

> In passato, l'unica soluzione a questi problemi era un foglio di stile separato con sostituzioni per diversi sistemi di scrittura. Tuttavia, più recentemente i CSS hanno acquisito [proprietà e valori *logici*](https://developer.mozilla.org/it/docs/Web/CSS/CSS_Logical_Properties) che funzionano proprio come i loro omologhi *fisici* ma sono sensibili alla direzionalità del loro contesto. Ad esempio, invece di `width` potremmo scrivere `inline-size`, e invece di `left` potremmo usare [`inset-inline`](https://developer.mozilla.org/it/docs/Web/CSS/inset-inline). Oltre alle *proprietà* logiche, ci sono anche parole *chiave* logiche, come `float: inline-start` invece di `float: left`.

Reviewer: ```suggestion
In passato, l'unica soluzione a questi problemi era un foglio di stile separato con sostituzioni per diversi sistemi di scrittura. Tuttavia, più recentemente i CSS hanno acquisito [proprietà e valori *logici*](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_Logical_Properties) che funzionano proprio come i loro omologhi *fisici* ma sono sensibili alla direzionalità del loro contesto. Ad esempio, invece di `width` potremmo scrivere `inline-size`, e invece di `left` potremmo usare [`inset-inline`](https://developer.mozilla.org/en-US/docs/Web/CSS/inset-inline). Oltre alle *proprietà* logiche, ci sono anche parole *chiave* logiche, come `float: inline-start` invece di `float: left`.
```

  chart_url="https://docs.google.com/spreadsheets/d/e/2PACX-1vRpe_HsNGpekn6YZV9k6QGmcZPxalqnDrL7DrDY-7X65RZEf_-aGfWuEvhk-yWV83ctIceE1bppCLpj/pubchart?oid=744523431&format=interactive",
  sheets_gid="1466863581",
  sql_file="vendor_prefix_pseudo_elements.sql"
) }}

Quasi tutto l'uso di funzioni con prefisso (98%) è per specificare i gradienti, anche se [questo non è stato necessario dal 2014](https://caniuse.com/css-gradients). Il più popolare di questi è `-webkit-linear-gradient()` usato in oltre un quarto delle pagine esaminate. Il restante <2% è principalmente per calc, [per il quale non è stato necessario un prefisso dal 2013](https://caniuse.com/calc).

{{ figure_markup(
  caption="Percentuale di funzioni gradiente in tutte le occorrenze di funzioni con prefisso fornitore nelle pagine per dispositivi mobile.",
  content="98.22%",
  classes="big-number",
  sheets_gid="1586213539",
  sql_file="vendor_prefix_functions.sql"
) }}

> L'utilizzo delle funzionalità multimediale con prefisso è complessivamente inferiore, con la più popolare, `-webkit-min-pixel-ratio`, utilizzata nel 13% delle pagine per rilevare i display "Retina". La corrispondente funzionalità multimediale standard, [`resolution`](https://developer.mozilla.org/it/docs/Web/CSS/@media/resolution) l'ha finalmente superata in popolarità ed è usata nel 22% degli pagine per dispositivi mobile e il 15% di pagine desktop.
> 
> Complessivamente, `-*-min-pixel-ratio` comprende tre quarti delle funzionalità multimediali con prefisso su desktop e circa la metà su dispositivi mobile. La ragione della differenza non è il ridotto utilizzo dei dispositivi mobile, ma il fatto che un'altra funzione multimediale con prefisso, `-*-high-contrast`, è molto più popolare sui dispositivi mobile e rappresenta quasi tutta l'altra metà delle funzioni multimediali con prefisso, ma solo il 18% su desktop. La funzione multimediale standard corrispondente, [forced-colors](https://developer.mozilla.org/it/docs/Web/CSS/@media/forced-colors) è ancora sperimentale in Chrome e Firefox.

Reviewer: ```suggestion
L'utilizzo delle funzionalità multimediale con prefisso è complessivamente inferiore, con la più popolare, `-webkit-min-pixel-ratio`, utilizzata nel 13% delle pagine per rilevare i display "Retina". La corrispondente funzionalità multimediale standard, [`resolution`](https://developer.mozilla.org/en-US/docs/Web/CSS/@media/resolution) l'ha finalmente superata in popolarità ed è usata nel 22% degli pagine per dispositivi mobile e il 15% di pagine desktop.

Complessivamente, `-*-min-pixel-ratio` comprende tre quarti delle funzionalità multimediali con prefisso su desktop e circa la metà su dispositivi mobile. La ragione della differenza non è il ridotto utilizzo dei dispositivi mobile, ma il fatto che un'altra funzione multimediale con prefisso, `-*-high-contrast`, è molto più popolare sui dispositivi mobile e rappresenta quasi tutta l'altra metà delle funzioni multimediali con prefisso, ma solo il 18% su desktop. La funzione multimediale standard corrispondente, [forced-colors](https://developer.mozilla.org/en-US/docs/Web/CSS/@media/forced-colors) è ancora sperimentale in Chrome e Firefox.
```

{{ figure_markup(
  image="vendor-prefixed-media.png",
  caption="Popolarità relativa delle funzionalità multimediali con prefisso del fornitore come percentuale di occorrenze sulle pagine per dispositivi mobile.",
  description="Grafico a torta della popolarità relativa delle funzionalità multimediali con prefisso del fornitore come percentuale di occorrenze sulle pagine per dispositivi mobile. `min-device-pixel-ratio` e `high-contrast` costituiscono ciascuno il 47% delle occorrenze, `transform-3d` al 5% e le restanti caratteristiche inferiori all'1% sono `device-pixel-ratio`, `max-device-pixel-ratio`, e altre funzionalità.",
  chart_url="https://docs.google.com/spreadsheets/d/e/2PACX-1vRpe_HsNGpekn6YZV9k6QGmcZPxalqnDrL7DrDY-7X65RZEf_-aGfWuEvhk-yWV83ctIceE1bppCLpj/pubchart?oid=1940027848&format=interactive",
  sheets_gid="1192245087",
  sql_file="vendor_prefix_media.sql"
) }}

### Le funzionalità delle query

> Le funzionalità delle query ([@supports](https://developer.mozilla.org/it/docs/Web/CSS/@supports)) stanno guadagnando terreno negli ultimi anni e sono state utilizzate nel 39% dei pagine, in notevole aumento rispetto al 30% dello scorso anno.

Reviewer: ```suggestion
Le funzionalità delle query ([@supports](https://developer.mozilla.org/en-US/docs/Web/CSS/@supports)) stanno guadagnando terreno negli ultimi anni e sono state utilizzate nel 39% dei pagine, in notevole aumento rispetto al 30% dello scorso anno.
```

{{ figure_markup(
  image="margin-padding-shorthand-vs-longhand.png",
  caption="Confronto dell'utilizzo delle shorthand `margin` e `padding` e delle loro forme complete.",
  description="Il grafico a barre che mostra `padding` è del 93% su desktop, 94% su dispositivi mobile, `margin` è rispettivamente 93% e 93%, `margin-left` è 91% e 92%, `margin-top` è 90% e 91 %, `margin-right` è 90% e 91%, `margin-bottom` è 90% e 91%, `padding-left` è 90% e 90%, `padding-top` è 88% e 89%, `padding-bottom` è 88% e 89% e `padding-right` è 87% e 88%.",
  chart_url="https://docs.google.com/spreadsheets/d/e/2PACX-1vRpe_HsNGpekn6YZV9k6QGmcZPxalqnDrL7DrDY-7X65RZEf_-aGfWuEvhk-yWV83ctIceE1bppCLpj/pubchart?oid=804317202&format=interactive",
  sheets_gid="1513860089",
  sql_file="all_properties.sql"
) }}

#### Flex

> Quasi tutte le proprietà `flex`, `flex-*` sono molto usate e appaiono nel 30-60% delle pagine. Tuttavia, sia `flex-wrap` che `flex-direction` sono usati molto più della loro shorthand, `flex-flow`. Quando viene utilizzato `flex-flow`, viene utilizzato con due valori, ovvero come un modo più breve per impostare entrambe le forme complete. Nonostante gli [elaborati valori predefiniti ragionevoli](https://developer.mozilla.org/it/docs/Web/CSS/flex#Syntax:~:text=The%20flex%20property%20may%20be%20specified%20using%20one%2C%20two%2C%20or%20three%20values) per l'uso di `flex` con uno o due valori, circa il 90% dell'utilizzo consiste nella sintassi dei 3 valori, impostando esplicitamente tutti e tre delle sue forme complete.

Reviewer: ```suggestion
Quasi tutte le proprietà `flex`, `flex-*` sono molto usate e appaiono nel 30-60% delle pagine. Tuttavia, sia `flex-wrap` che `flex-direction` sono usati molto più della loro shorthand, `flex-flow`. Quando viene utilizzato `flex-flow`, viene utilizzato con due valori, ovvero come un modo più breve per impostare entrambe le forme complete. Nonostante gli [elaborati valori predefiniti ragionevoli](https://developer.mozilla.org/it/docs/Web/CSS/flex#syntax) per l'uso di `flex` con uno o due valori, circa il 90% dell'utilizzo consiste nella sintassi dei 3 valori, impostando esplicitamente tutti e tre delle sue forme complete.
```

- Il 37% di loro era una forma alterata di una proprietà con prefisso (ad es. `webkit-Transition` o `-transition`)
- Il 43% era una forma non prefissata di una proprietà che esiste solo con un prefisso (ad esempio `font-smoothing`, che appariva su 384K siti web), probabilmente inclusa per compatibilità sotto l'errato presupposto che sia standard, o a causa di un pio desiderio che diventerà standard.
- Un errore di battitura che ha trovato la sua strada in una libreria popolare. Attraverso questa analisi, abbiamo scoperto che la proprietà `white-wpace` era presente in 234.027 siti web. Ci sono troppi siti Web perché lo stesso errore di battitura si sia verificato in modo organico, quindi abbiamo deciso di esaminarlo. Ed ecco, [si scopre](https://twitter.com/rick_viscomi/status/1326739379533000704) era il widget di Facebook! La correzione è già in atto.

> - E un'altra stranezza: la proprietà `font-rendering` appare su 2.575 pagine. Tuttavia, non possiamo trovare prove dell'esistenza di tale proprietà, con o senza un prefisso. C'è il non standard [`-webkit-font-smoothing`](https://medium.com/better-programming/improving-font-rendering-with-css-3383fc358cbc) che è molto popolare, comparendo in 3 milioni di siti web, o circa il 49% delle pagine, ma `font-rendering` non è sufficientemente vicino per essere un errore di ortografia. C'è [`text-rendering`](https://developer.mozilla.org/it/docs/Web/CSS/text-rendering) che viene utilizzato in circa 100K di siti web, quindi è concepibile che 2.5K tutti gli sviluppatori abbiano erroneamente scambiato `font-smoothing` e `text-rendering`.

Reviewer: ```suggestion
- E un'altra stranezza: la proprietà `font-rendering` appare su 2.575 pagine. Tuttavia, non possiamo trovare prove dell'esistenza di tale proprietà, con o senza un prefisso. C'è il non standard [`-webkit-font-smoothing`](https://medium.com/better-programming/improving-font-rendering-with-css-3383fc358cbc) che è molto popolare, comparendo in 3 milioni di siti web, o circa il 49% delle pagine, ma `font-rendering` non è sufficientemente vicino per essere un errore di ortografia. C'è [`text-rendering`](https://developer.mozilla.org/en-US/docs/Web/CSS/text-rendering) che viene utilizzato in circa 100K di siti web, quindi è concepibile che 2.5K tutti gli sviluppatori abbiano erroneamente scambiato `font-smoothing` e `text-rendering`.
```

  sheets_gid="1743048352",
  sql_file="meta_important_adoption.sql"
) }}

Cos'è che gli sviluppatori sono così desiderosi di sovrascrivere? Abbiamo esaminato la suddivisione per proprietà e abbiamo scoperto che quasi l'80% delle pagine utilizza `!important` con la proprietà `display`. È una strategia comune applicare `display: none !important` per nascondere il contenuto nelle classi helper per sovrascrivere i CSS esistenti che usano `display` per definire una modalità di layout. Questo è un effetto collaterale di quello che, col senno di poi, era un difetto nei CSS. Ha combinato tre funzionalità ortogonali in una: la modalità di layout interno, il comportamento del flusso e lo stato di visibilità sono tutti controllati dalla proprietà `display`. Sono stati fatti sforzi per separare questi valori in singole parole chiave `display` in modo che possano essere regolate individualmente tramite proprietà personalizzate, ma per il momento [il supporto del browser è praticamente inesistente.](https://caniuse.com/mdn-css_properties_display_multi-keyword_values)

{{ figure_markup(
  image="important-top-properties.png",

>   alt="Le prime proprietà `!important` in base alla percentuale di pagine.",

Reviewer: Alt text should be without HMTL or markdown:

```suggestion
  alt="Le prime proprietà !important in base alla percentuale di pagine.",
```
